clash-lib/src/Clash/Core/Util.hs

```diff
@@ -628,3 +783,61 @@ mkUniqInternalId (supply,inScope) (nm, ty) =
   (u,supply') = freshId supply
   v           = mkId ty (mkUnsafeInternalName nm u)
   v'          = uniqAway inScope v
+
+
+-- | Same as @dataConInstArgTys@, but it tries to compute existentials too,
+-- hence the extra argument @TyConMap@. WARNING: It will return the types
+-- of non-existentials only
+dataConInstArgTysE
+  :: HasCallStack
+  => InScopeSet
+  -> TyConMap
+  -> DataCon
+  -> [Type]
+  -> Maybe [Type]
+dataConInstArgTysE is0 tcm (MkData { dcArgTys, dcExtTyVars, dcUnivTyVars }) inst_tys =
+  -- TODO: Check if all existentials were solved (they should be, or the wouldn't have
+  -- TODO: been solved in the caseElemExistentials transformation)
+  let is1   = extendInScopeSetList is0 dcExtTyVars in
+  let is2   = unionInScope is1 (mkInScopeSet (tyFVsOfTypes inst_tys)) in
+  let subst = extendTvSubstList (mkSubst is2) (zip dcUnivTyVars inst_tys) in
+  go
+    (substGlobalsInExistentials is0 dcExtTyVars (zip dcUnivTyVars inst_tys))
+    (map (substTy subst) dcArgTys)
+
+ where
+  go
+    :: [TyVar]
+    -- ^ Existentials
+    -> [Type]
+    -- ^ Type arguments
+    -> Maybe [Type]
+    -- ^ Maybe (solved existentials, types of non-existentials)
+  go exts0 args0 =
+    let eqs = catMaybes (map (typeEq tcm) args0) in
+    case solveFirstNonAbsurd eqs of
+      Just (tyVar, solution1) ->
+        go exts1 args1
+        where
+          solution2 = LitTy (NumTy solution1)
+          exts1 = substInExistentials is0 exts0 (tyVar, solution2)
+          is2   = extendInScopeSetList is0 exts1
+          subst = extendTvSubst (mkSubst is2) tyVar solution2
+          args1 = map (substTy subst) args0
+      Nothing ->
+        Just args0
+
+
+-- | Given a DataCon and a list of types, the type variables of the DataCon
+-- type are substituted for the list of types. The argument types are returned.
+--
+-- The list of types should be equal to the number of type variables, otherwise
+-- @Nothing@ is returned.
+dataConInstArgTys :: DataCon -> [Type] -> Maybe [Type]
+dataConInstArgTys (MkData { dcArgTys, dcUnivTyVars, dcExtTyVars }) inst_tys =
+  let tyvars = dcUnivTyVars ++ dcExtTyVars in
+  if length tyvars == length inst_tys then
+    -- Notice it is safe to use substTyWith here, as ALL variables are replaced
```

Comment:
It's safe to use subsTyWith here because:

1. The [Note] subst invaration says: FVs should be inScopeSet or in the range of the subst
2. The FVs of dcArgTys can only be tyvars
3. The range of the substitution is tyvars, thus satisfying the invariant of 1.